### ISSUE TITLE
Byte Arrays

### ISSUE THREAD

**Issue by mcf-rocks** [near/borsh#103]
```
use borsh::{BorshSerialize, BorshDeserialize};

#[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
struct B {
    x: [u8; 20],
    y: [u8; 100],
    z: String,
}

fn test_simple_struct() {
    let b = B {
        x: [0; 20],
        y: [0; 100],
        z: "liber primus".to_string(),
    };
    let encoded_b = b.try_to_vec().unwrap();
    let decoded_b = B::try_from_slice(&encoded_b).unwrap();
    assert_eq!(b, decoded_b);
}

fn main() {
    test_simple_struct();
}
```

```
~/BORSH/borsh-test/src$ cargo run
   Compiling borsh-test v0.1.0 (/Users/mrsmith/BORSH/borsh-test)
error[E0277]: the trait bound `[u8; 100]: borsh::BorshDeserialize` is not satisfied
 --> src/main.rs:4:26
  |
4 | #[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
  |                          ^^^^^^^^^^^^^^^^ the trait `borsh::BorshDeserialize` is not implemented for `[u8; 100]`
  |
  = help: the following implementations were found:
            <[T; 0] as borsh::BorshDeserialize>
            <[T; 1024] as borsh::BorshDeserialize>
            <[T; 10] as borsh::BorshDeserialize>
            <[T; 11] as borsh::BorshDeserialize>
          and 36 others
  = help: see issue #48214
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `[u8; 100]: borsh::BorshSerialize` is not satisfied
 --> src/main.rs:4:10
  |
4 | #[derive(BorshSerialize, BorshDeserialize, PartialEq, Debug)]
  |          ^^^^^^^^^^^^^^ the trait `borsh::BorshSerialize` is not implemented for `[u8; 100]`
  |
  = help: the following implementations were found:
            <[T; 0] as borsh::BorshSerialize>
            <[T; 1024] as borsh::BorshSerialize>
            <[T; 10] as borsh::BorshSerialize>
            <[T; 11] as borsh::BorshSerialize>
          and 37 others
  = help: see issue #48214
  = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
error: could not compile `borsh-test`.

To learn more, run the command again with --verbose.
```

**Comment by mcf-rocks** [comment]
Seems to be ok with the 20, but not the 100 ??

**Comment by prographo** [comment]
Are only some array lengths implemented? 

https://github.com/vidhunv1/borsh/blob/3e0fc3d328c67851ec79aef83e0d9c10f3616275/borsh-rs/borsh/src/ser/mod.rs#L347

**Comment by willemneal** [comment]
I think @prographo is correct given:
```
^ the trait `borsh::BorshDeserialize` is not implemented for `[u8; 100]`
```

@nearmax would know more about this.  Max is it possible to make this a range instead of a list of numbers?  And why is it not just any array type with length?

**Comment by nearmax** [comment]
Array length cannot be a type parameter in Rust, so derive traits like this one usually just implement trait for some number of array lengths. This specific sequence of lengths was taken from serde AFAIR. Until Rust lifts this limitation there is no way to define it for a general array.

**Comment by SuperFluffy** [comment]
Just a heads up in case you have not yet seen this: `min_const_generics` is on route to be stabilized in 1.50: https://github.com/rust-lang/rust/pull/79135

If my reading is correct, this will allow implementing `Borsh{De,}Serialize` for `[T; const N: usize]`.

**Comment by frol** [comment]
`const-generics` feature is implemented and released as of borsh 0.9.1+ (just for the reference: https://github.com/near/borsh/issues/36#issuecomment-882934079)
